packages/vite/src/node/server/moduleGraph.ts

```diff
@@ -150,15 +150,24 @@ export class ModuleGraph {
     const prevImports = mod.importedModules
     const nextImports = (mod.importedModules = new Set())
     let noLongerImported: Set<ModuleNode> | undefined
+    let resolvePromises = []
     // update import graph
     for (const imported of importedModules) {
-      const dep =
-        typeof imported === 'string'
-          ? await this.ensureEntryFromUrl(imported, ssr)
-          : imported
-      dep.importers.add(mod)
-      nextImports.add(dep)
+      if (typeof imported === 'string') {
+        resolvePromises.push(
+          this.ensureEntryFromUrl(imported, ssr).then((dep) => {
+            dep.importers.add(mod)
+            nextImports.add(dep)
+          }),
+        )
+      } else {
+        imported.importers.add(mod)
+        nextImports.add(imported)
+      }
     }
+
+    resolvePromises.length && (await Promise.all(resolvePromises))
```

Comment:
Tiny nit: I think doing
```suggestion
    if (resolvePromises.length) await Promise.all(resolvePromises)
```
would be more consistent in the codebase.